### PR TITLE
Study card reapplication

### PR DESCRIPTION
**Pull request by Nunien**
* reapply studycard button in solr
* show dicom metadata button in dataset details

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java [hunk 1]
```diff
@@ -107,7 +117,7 @@
 	 */
 	public void downloadDicomFilesForURLs(final List<URL> urls, final File workFolder, String subjectName) throws IOException, MessagingException {
 		for (Iterator iterator = urls.iterator(); iterator.hasNext();) {
-			String url = ((URL) iterator.next()).toString();
+			String url = ((URL) iterator.next()).toString() + "&";
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java:120]: What is it for ? :)

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java [hunk 2]
```diff
@@ -139,6 +149,17 @@ public void downloadDicomFilesForURLs(final List<URL> urls, final File workFolde
 			}
 		}
 	}
+	
+	public String downloadDicomMetadataForURL(final URL url) throws IOException, MessagingException {
+		if (url != null) {
+			String urlStr = url.toString();
+			if (urlStr.contains(WADO_REQUEST_STUDY_WADO_URI)) urlStr = wadoURItoWadoRS(urlStr);
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java:156]: This is based on the supposition that the PACS accepts REST requests, but it's not always the case no? Am I wrong @michaelkain ?
Maybe you should test
if (dicomWeb) { REST } else { WadoURI}

File: shanoir-ng-front/src/app/datasets/dataset/dataset.component.html [hunk 3]
```diff
@@ -59,6 +59,7 @@ <h2 class="header command-zone"i18n="Create dataset|Title@@datasetDetailCreateTi
 				<button *ngIf="mode=='view' && hasDownloadRight && dataset.type != 'Eeg'" [disabled]="downloading" type="button" class="right-icon" i18n="Buttons|Download button label@@downloadButton" (click)="download('dcm')">Download Dicom<i class="fas fa-download"></i></button>
 				<button *ngIf="mode=='view' && hasDownloadRight && dataset.type != 'Eeg'" [disabled]="downloading" type="button" class="right-icon" i18n="Buttons|Download button label@@downloadButton" (click)="download('nii')">Download Nifti<i class="fas fa-download"></i></button>
 				<button *ngIf="mode=='view' && hasDownloadRight && dataset.type == 'Eeg'" [disabled]="downloading" type="button" class="right-icon" i18n="Buttons|Download button label@@downloadButton" (click)="download('eeg')">Download EEG<i class="fas fa-download"></i></button>
+				<button *ngIf="mode=='view' && hasDownloadRight" type="button" class="right-icon" (click)="seeDicomMetadata()">Show DICOM metadata<i class="fas fa-file-code"></i></button>
```
Comment (jcomedouteau) [inline shanoir-ng-front/src/app/datasets/dataset/dataset.component.html:62]: This should not appear on EEG data I guess

File: shanoir-ng-front/src/app/datasets/dataset/metadata/metadata.component.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            breadcrumbsService: BreadcrumbsService,
+            private location: Location) {
+
+        breadcrumbsService.nameStep('Dicom metadata');
+        this.loadMetadata();
+    }
+
+    private loadMetadata() {
+        const id: number = +this.activatedRoute.snapshot.params['id'];
+        Promise.all([this.datasetService.downloadDicomMetadata(id), this.dicomService.getDicomTags()]).then(([data, tags]) => {
```
Comment (jcomedouteau) [inline shanoir-ng-front/src/app/datasets/dataset/metadata/metadata.component.ts:46]: If the response is a 204, is the message "no metadata" displayed ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java (hunk 1 above)
Comment (Nunien) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java]: I don't know I removed it and it seems to work

File: shanoir-ng-front/src/app/datasets/dataset/dataset.component.html (hunk 3 above)
Comment (Nunien) [reply shanoir-ng-front/src/app/datasets/dataset/dataset.component.html]: I added a check

File: shanoir-ng-front/src/app/datasets/dataset/metadata/metadata.component.ts (hunk 4 above)
Comment (Nunien) [reply shanoir-ng-front/src/app/datasets/dataset/metadata/metadata.component.ts:46]: I did that

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetUtils.java [hunk 5]
```diff
@@ -33,4 +49,69 @@ public static void getDatasetFilePathURLs(final Dataset dataset, List<URL> pathU
...
+				break;
+			case StatisticalDataset.datasetType:
+				dataset = new StatisticalDataset();
+				originMetadata.setDatasetModalityType(DatasetModalityType.GENERIC_DATASET);
+				break;
+			case TemplateDataset.datasetType:
+				dataset = new TemplateDataset();
+				originMetadata.setDatasetModalityType(DatasetModalityType.GENERIC_DATASET);
+				break;
+			default:
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/DatasetUtils.java:116]: done, and missing usage of originMetadata corrected, as setter never called.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java (hunk 2 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/download/WADODownloaderService.java:174]: The base for this PR is a new PACS. Without it neither OHIF viewer, nor re-apply studycards nor quality control will work. We have not yet changed to store paths in wado-rs, so it should be fine.

**Review by michaelkain — APPROVED** [review]

**Comment by michaelkain** [comment]
I require this PR to start developing in the backend for quality control. The study cards mechanism and its re-application are a function in relation with the quality control extension, e.g. qc might be applied on elements changed by the study cards. Use the harmonization power of study cards and then check the quality on the harmonized data (possible, but not mandatory).

**Comment by michaelkain** [comment]
In my local tests the re-application of a study card does not work on one dataset. As I require this PR for quality control I will merge it and check what piece is missing to re-apply study cards. The GUI displays a DONE message, but the dataset name is not changed, even if part of the rule: to investigate. Imports and everything else works fine.

**Pull request merged**